// typedjson/lib/esm/json-map-member.js
import { isReflectMetadataSupported, logError, MISSING_REFLECT_CONF_MSG, nameof } from './helpers';
import { injectMetadataInformation } from './metadata';
import { extractOptionBase } from './options-base';
import { ensureTypeThunk, MapT } from './type-descriptor';
export function jsonMapMember(maybeKeyThunk, maybeValueThunk, options = {}) {
    return (target, propKey) => {
        var _a;
        const decoratorName = `@jsonMapMember on ${nameof(target.constructor)}.${String(propKey)}`;
        const keyThunk = ensureTypeThunk(maybeKeyThunk, decoratorName);
        const valueThunk = ensureTypeThunk(maybeValueThunk, decoratorName);
        const reflectedType = isReflectMetadataSupported
            ? Reflect.getMetadata('design:type', target, propKey)
            : null;
        if (reflectedType != null && reflectedType !== Map && reflectedType !== Object) {
            logError(`${decoratorName}: property is not a Map. ${MISSING_REFLECT_CONF_MSG}`);
            return;
        }
        injectMetadataInformation(target, propKey, {
            type: () => MapT(keyThunk(), valueThunk(), { shape: options.shape }),
            emitDefaultValue: options.emitDefaultValue,
            isRequired: options.isRequired,
            options: extractOptionBase(options),
            key: propKey.toString(),
            name: (_a = options.name) !== null && _a !== void 0 ? _a : propKey.toString(),
            deserializer: options.deserializer,
            serializer: options.serializer,
        });
    };
}
