// typedjson/lib/esm/json-set-member.js
import { isReflectMetadataSupported, logError, MISSING_REFLECT_CONF_MSG, nameof } from './helpers';
import { injectMetadataInformation } from './metadata';
import { extractOptionBase } from './options-base';
import { ensureTypeThunk, SetT } from './type-descriptor';
export function jsonSetMember(maybeTypeThunk, options = {}) {
    return (target, propKey) => {
        var _a;
        const decoratorName = `@jsonSetMember on ${nameof(target.constructor)}.${String(propKey)}`;
        const typeThunk = ensureTypeThunk(maybeTypeThunk, decoratorName);
        const reflectedType = isReflectMetadataSupported
            ? Reflect.getMetadata('design:type', target, propKey)
            : null;
        if (reflectedType != null && reflectedType !== Set && reflectedType !== Object) {
            logError(`${decoratorName}: property is not a Set. ${MISSING_REFLECT_CONF_MSG}`);
            return;
        }
        injectMetadataInformation(target, propKey, {
            type: () => SetT(typeThunk()),
            emitDefaultValue: options.emitDefaultValue,
            isRequired: options.isRequired,
            options: extractOptionBase(options),
            key: propKey.toString(),
            name: (_a = options.name) !== null && _a !== void 0 ? _a : propKey.toString(),
            deserializer: options.deserializer,
            serializer: options.serializer,
        });
    };
}
