// unibeautify/dist/src/LanguageManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
class LanguageManager {
    constructor(languages) {
        this.languages = languages;
    }
    getLanguage(data) {
        const langs = this.findLanguages({
            atomGrammar: data.atomGrammar,
            extension: data.fileExtension,
            name: data.languageName,
            sublimeSyntax: data.sublimeSyntax,
        });
        return langs.length > 0 ? langs[0] : null;
    }
    /**
    Find and return the appropriate Languages that match any of the given filter criteria.
    An empty array will be returned if there are no matches.
  
    Precedence:
    - name
    - namespace
    - extension
    - atomGrammar
    - sublimeSyntax
    - vscodeLanguage
    */
    findLanguages(query) {
        const filters = {
            name: query.name,
            namespace: query.namespace,
            // tslint:disable-next-line:object-literal-sort-keys
            extensions: query.extension,
            atomGrammars: query.atomGrammar,
            sublimeSyntaxes: query.sublimeSyntax,
            vscodeLanguages: query.vscodeLanguage,
        };
        const langs = utils_1.filterMultiCriteria(this.languages, filters);
        return utils_1.unique(langs);
    }
    /**
    Get a shallow copy of the languages currently loaded.
    */
    getLoadedLanguages() {
        return this.languages.slice();
    }
    /**
    Load a Language
    */
    loadLanguage(language) {
        this.languages.push(language);
        return this;
    }
    /**
    Load multiple Languages
    */
    loadLanguages(languages) {
        this.languages.push(...languages);
        return this;
    }
}
exports.LanguageManager = LanguageManager;
